### PR TITLE
Added specification enhance_gluster_cluster_sync_logic.adoc

### PR DESCRIPTION
**Pull request by shtripat**
tendrl-bug-id: Tendrl/specifications#167
Signed-off-by: Shubhendu <shtripat@redhat.com>

### CODE REVIEW

**Comment by shtripat** [comment]
@Tendrl/tendrl-core @atinmu @samikshan plz review.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (nnDarshan) [inline specs/enhance_gluster_cluster_sync_logic.adoc]: We might have to mark the brick that takes the role of arbiter as well. This would involve modification to brick object. Can you please add some details around this.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (nnDarshan) [inline specs/enhance_gluster_cluster_sync_logic.adoc]: Since some snapshot details like snap hard/soft limit are available as part of `gluster get-state --volume-options` (I am guessing the options would be available by some command like this). How are we planning to club the details from these 2 commands. Since the above mentioned command is something that we should invoke only once. Can you please add some details around the strategy to collect those details.

Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: I feel snapshot quota options are volume options and they can still be maintained under volume options entity and not as part of volume snapshots. Volume snapshots entity maintains only the list of snapshot (if any) for the volume. Hope this makes sense?

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: Sure. We can have a flag `arbiter_brick` set as boolean at brick level. Will add that.

Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: In `get-state` output we do get the `arbiter_count = 1` for an arbiter volume but there is no marking of the arbiter brick as such. A sample output is as below

```
[Volumes]
Volume1.name: arb-vol
Volume1.id: 543fd98b-b876-4027-826a-c051761b5b40
Volume1.type: Replicate
Volume1.transport_type: tcp
Volume1.status: Started
Volume1.brickcount: 3
Volume1.Brick1.path: dhcp43-118.lab.eng.blr.redhat.com:/root/gluster_bricks/vol3_b1
Volume1.Brick1.hostname: dhcp43-118.lab.eng.blr.redhat.com
Volume1.Brick2.path: dhcp43-116.lab.eng.blr.redhat.com:/root/gluster_bricks/vol3_b2
Volume1.Brick2.hostname: dhcp43-116.lab.eng.blr.redhat.com
Volume1.Brick3.path: dhcp43-126.lab.eng.blr.redhat.com:/root/gluster_bricks/vol3_b3
Volume1.Brick3.hostname: dhcp43-126.lab.eng.blr.redhat.com
Volume1.Brick3.port: 49152
Volume1.Brick3.rdma_port: 0
Volume1.Brick3.status: Started
Volume1.Brick3.signedin: True
Volume1.snap_count: 0
Volume1.stripe_count: 1
Volume1.replica_count: 3
Volume1.subvol_count: 1
Volume1.arbiter_count: 1
Volume1.disperse_count: 0
Volume1.redundancy_count: 0
Volume1.quorum_status: not_applicable
Volume1.snapd_svc.online_status: Offline
Volume1.snapd_svc.inited: True
Volume1.rebalance.id: 00000000-0000-0000-0000-000000000000
Volume1.rebalance.status: not_started
Volume1.rebalance.failures: 0
Volume1.rebalance.skipped: 0
Volume1.rebalance.lookedup: 0
Volume1.rebalance.files: 0
Volume1.rebalance.data: 0Bytes
Volume1.options.transport.address-family: inet
Volume1.options.performance.readdir-ahead: on
Volume1.options.nfs.disable: on
```

@samikshan can we add that detail here?

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (r0h4n) [inline specs/enhance_gluster_cluster_sync_logic.adoc]: Each volume{n},  should be synced on a new greenlet

**Review by r0h4n — CHANGES_REQUESTED** [review]

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: Ack .  Will take care while code changes....

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (atinmu) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: arbiter brick will be always the last brick (3rd to be precise) in the chain.

Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: As discussed today it would be taken care in `get-state` output. The github issue created fro the same is at https://github.com/gluster/glusterfs/issues/278

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (cybernth) [inline specs/enhance_gluster_cluster_sync_logic.adoc:7]: By implementing this spec, I assume that we are removing the dependency with gstatus? if that is the case we need to account for the spec file changes as well

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (cybernth) [inline specs/enhance_gluster_cluster_sync_logic.adoc:38]: How we are planning to calculate volume status? I think `gluster volume status` gives the info whether the volume is started or not? gstatus gives states like `degraded` etc. We need to have something similar

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (cybernth) [inline specs/enhance_gluster_cluster_sync_logic.adoc:42]: Can you add sample outputs for `get-state detail` and `get-state volumeoptions` commands

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (cybernth) [inline specs/enhance_gluster_cluster_sync_logic.adoc]: What about :
1. Heal information
2. What are the different volume types we support now? do we support all?

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (cybernth) [inline specs/enhance_gluster_cluster_sync_logic.adoc]: please add this info

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (cybernth) [inline specs/enhance_gluster_cluster_sync_logic.adoc]: Why do you think these APIs are required?. We are not going to do day1,day2 now

**Review by cybernth — CHANGES_REQUESTED** [review]

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc:7]: Yes we remove the dependency of `gstatus` with these changes. Will add the spec change as well.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc:38]: Volume status is something which is returned in `get-state` and we will use that as earlier. Volume state will be calculated using the same logic which `gstatus` had to form the cluster status value. Code changes PR would give more idea (actually the logic is copied from `gstatus`).

Do you need volume status as degraded etc? Earlier also we were using only `get-state` output for volume status. The volume state was used by `gstatus` only for deciding the cluster status (healthy/unhealthy)

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc:42]: `get-state detail` output is there at bottom of the spec. Will add a sample of `get-state volumeoptions` as well.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: Everything which is available by `get-state detail` and `get-state volumeoptions` is covered as part of this spec. As part of this we have `estimated_time_left` for rebalance though which is calculated based on suggestion by gluster team (maximum of `time_left` reported from different nodes for rebalance for a volume).

Should we run `heal info` as well for each volume and sync details? I understand heal should be enabled/disabled also at volume level.

For cluster status and utilization all the volume types are supported actually and the logic is copied from `gstatus`

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: Done. This would be added once we have geo-rep detail available in `get-state` which is planned.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: Listing API should be available I feel.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (brainfunked) [inline specs/enhance_gluster_cluster_sync_logic.adoc:47]: What are the details and how would they be calculated? The logic needs to be documented in the spec and verified before implementing in code.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (brainfunked) [inline specs/enhance_gluster_cluster_sync_logic.adoc:57]: Are the details readily available in the `get-state` output or does some logic need to be applied?

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (brainfunked) [inline specs/enhance_gluster_cluster_sync_logic.adoc]: Please document the aggregation strategy.

Also, I think there are certain pitfalls in using the client connections provided by gluster, such as no differentiation between actual clients and other processes. This needs to be documented clearly in the spec as it would be necessary for the documentation of this feature.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (brainfunked) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: @shtripat please document these details in the spec itself.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (brainfunked) [inline specs/enhance_gluster_cluster_sync_logic.adoc]: Please document the aggregation logic.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (brainfunked) [inline specs/enhance_gluster_cluster_sync_logic.adoc]: What's the aggregation logic?

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (brainfunked) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: Document in the spec please.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (brainfunked) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: @atinmu what's your take on this?

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (brainfunked) [inline specs/enhance_gluster_cluster_sync_logic.adoc:406]: I believe there are some changes to the `get-state` that need to be made in gluster upstream. All the gluster upstream issues tracking these changes need to be listed as dependencies. @atinmu may be able to help with the list of these issues.

**Review by brainfunked — CHANGES_REQUESTED** [review]

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc:47]: There is no logic involved in getting snapshot details from `get-state` output. All the snapshots list is available in `get-state` output and same would be captured and instances of class `Snapshot` (defined in data model section) would be created and persisted in central store.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc:57]: Gluster team is not yet ready with geo-rep details in `get-state` output. Once its available details of the structure of object etc and if any aggregations would be documented in this spec.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: Aggregation logic for client connections is simple. Simply add all the client connection for the participating bricks and that makes the value for volume.
Will add this details with notice that its as per guidelines from gluster team.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: Sure

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: Sure. would add details

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc:406]: Dependencies section does list the items required from gluster team and also the references section has link to glusterfs github issues link. What else is missing here?

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (cybernth) [reply specs/enhance_gluster_cluster_sync_logic.adoc:7]: @anmolbabu , FYI Please make sure collectd plugins doesn't have dependency to gstatus going forward

Comment (anmolbabu) [reply specs/enhance_gluster_cluster_sync_logic.adoc:7]: @nthomas-redhat Ack.. This has been done on existing plugins as in https://github.com/Tendrl/node-monitoring/pull/79 but yes that PR will eventually have to land on node-agent rather than node-monitoring(that its currently on)

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (cybernth) [reply specs/enhance_gluster_cluster_sync_logic.adoc:38]: Is there any specific reason why you maintain this separately? Can you list down the possible values for `volume state' and 'volume status' here? I think its better to have single state/status for volume

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (cybernth) [reply specs/enhance_gluster_cluster_sync_logic.adoc:42]: ack

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (cybernth) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: Add this only if there is a specific requirement

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc:38]: The status and state are two different things. Volume might be degraded due to some replicas not available or it could be partially available. But still volume status is `Started`. I feel it fine to have like this.
Volume state is something which dont maintain in central store for a volume. We use this state data to figure out the cluster's healthy/unhealthy status only. Still you feel details required in spec, I can add all possible values.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: Ack. @anivargi ^^^

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc:38]: @nthomas-redhat As discussed based on bricks status and type of the volume, this logic reports 4 volume states which `up`, `down`, `up(partial)` and `up(degraded)`. The same values can be updated as additional field `state` under volume with this logic.

The volume `state` aggregated based on above logic is more detailed and can be used to display in UI.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [inline specs/enhance_gluster_cluster_sync_logic.adoc:84]: @atinmu below is the pseudo code for calculation of volume and cluster utilization. The logic is simple copy from `gstatus` at the moment. Need your inputs to figure if its correct or some changes would be required in this logic.
May be if required you can add others from Gluster team to review the logic as well.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (cybernth) [reply specs/enhance_gluster_cluster_sync_logic.adoc:38]: ack

**Review by r0h4n — DISMISSED** [review]

**Review by cybernth — DISMISSED** [review]
Looks good

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (brainfunked) [reply specs/enhance_gluster_cluster_sync_logic.adoc:38]: Does gdeploy also report the `up (partial)` status? I know it returns `up (degraded)`. I'm asking because if this is a new status that we report, it'll need to be documented to tell the users what it actually means.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (brainfunked) [reply specs/enhance_gluster_cluster_sync_logic.adoc:57]: Is there a gluster issue we could link to here?

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (brainfunked) [inline specs/enhance_gluster_cluster_sync_logic.adoc]: s/a/an

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (brainfunked) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: The APIs will soon be required, most likely in the [Gluster Monitoring milestone](https://github.com/Tendrl/specifications/milestone/1). Though, I suggest that for now, we exclude the API details from this spec and file a separate spec or an issue for API endpoints for all the additional gluster state information we now include as part of this spec.

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (brainfunked) [reply specs/enhance_gluster_cluster_sync_logic.adoc:406]: Would be ideal to link to the gluster issues tracking these changes.

**Review by brainfunked — COMMENTED** [review]
Looks good for implementation, but I'd suggest addressing some minor concerns such as linking to gluster issues in this spec before merging it.

@r0h4n @nthomas-redhat please feel free to merge this spec, without awaiting an explicit ack from me, when the requested additions (if possible) are made to the document.

**Comment by shtripat** [comment]
>> Looks good for implementation, but I'd suggest addressing some minor concerns such as linking to
>> gluster issues in this spec before merging it.

The references section does have links to glusterfs github issues.

**Comment by shtripat** [comment]
>> Does gdeploy also report the up (partial) status? I know it returns up (degraded). I'm asking because if
>> this is a new status that we report, it'll need to be documented to tell the users what it actually means.

`up (partial)` and `up (degraded)` are two different states reported for volumes based on calculations. The pseudo code has the logic which decides in what situations the state would be `up (degraded)`

File: specs/enhance_gluster_cluster_sync_logic.adoc
Comment (shtripat) [reply specs/enhance_gluster_cluster_sync_logic.adoc]: Done

**Comment by shtripat** [comment]
>> The APIs will soon be required, most likely in the Gluster Monitoring milestone. Though, I suggest that
>> for now, we exclude the API details from this spec and file a separate spec or an issue for API endpoints
>> for all the additional gluster state information we now include as part of this spec.

Will remove the entry from this spec. @anivargi please do the needful from API side.

**Comment by shtripat** [comment]
>> Would be ideal to link to the gluster issues tracking these changes.

I have gluster github issues in references section. Will move here.

**Review by cybernth — APPROVED** [review]

**Review by r0h4n — APPROVED** [review]

**Pull request merged**
